### DISCUSSION TITLE
how to understand Tuple and TupleRow in Doris?

### DISCUSSION THREAD

**Discussion by englefly** [apache/doris#7233] [A - General / Q&A]
请多指教 :)
1. be/src/runtime/tuple.h
这里的tuple和table中的一行数据有什么关系？为什么这里的tuple里的slots是按数据类型大小排序的，而不是schema中table定义的字段排序的？

2. be/src/runtime/tuple_row.h
// A TupleRow encapsulates a contiguous sequence of Tuple pointers which
// together make up a row.
这里的row 和 table里的一行数据有什么关系？为什么是多个tuple，而不是多个字段？

**Comment by morningman** [comment]
Tuple 其实表示的是查询规划中，某个算子的一行。Tuple中的slot不是按大小排序的，这个你可能看错了。
TupleRow 包含1个或多个tuple。比如 一个 scannode 的tuplerow只有有一个 tuple，而一个join node 的tuplerow有左右两个子节点的tuple
